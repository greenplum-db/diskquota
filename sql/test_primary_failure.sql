CREATE SCHEMA ftsr;
SELECT diskquota.set_schema_quota('ftsr', '1 MB');
SET search_path TO ftsr;
create or replace language plpythonu;
--
-- pg_ctl:
--   datadir: data directory of process to target with `pg_ctl`
--   command: commands valid for `pg_ctl`
--   command_mode: modes valid for `pg_ctl -m`  
--
create or replace function pg_ctl(datadir text, command text, command_mode text default 'immediate')
returns text as $$
    import subprocess
    if command not in ('stop', 'restart'):
        return 'Invalid command input'

    cmd = 'pg_ctl -l postmaster.log -D %s ' % datadir
    cmd = cmd + '-W -m %s %s' % (command_mode, command)

    return subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '')
$$ language plpythonu;

create or replace function pg_recoverseg(datadir text, command text)
returns text as $$
    import subprocess
    cmd = 'gprecoverseg -%s -d %s; exit 0; ' % (command, datadir)
    return subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '')
$$ language plpythonu;

CREATE TABLE a(i int);
INSERT INTO a SELECT generate_series(1,100);
INSERT INTO a SELECT generate_series(1,100000);
SELECT pg_sleep(5);
-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);

-- now one of primary is down
select pg_ctl((select datadir from gp_segment_configuration c where c.role='p' and c.content=0), 'stop');

-- switch mirror to primary
select gp_request_fts_probe_scan();

-- check GPDB status
select content, preferred_role, role, status, mode from gp_segment_configuration where content = 0;

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);

-- increase quota
SELECT diskquota.set_schema_quota('ftsr', '200 MB');

-- pull up failed primary
-- start_ignore
select pg_recoverseg((select datadir from gp_segment_configuration c where c.role='p' and c.content=-1), 'a');
select pg_sleep(10);
select pg_recoverseg((select datadir from gp_segment_configuration c where c.role='p' and c.content=-1), 'ar');
select pg_sleep(15);
select pg_recoverseg((select datadir from gp_segment_configuration c where c.role='p' and c.content=-1), 'a');
select pg_sleep(10);
select pg_recoverseg((select datadir from gp_segment_configuration c where c.role='p' and c.content=-1), 'ar');
select pg_sleep(10);
-- check GPDB status
select content, preferred_role, role, status, mode from gp_segment_configuration where content = 0;
SELECT pg_sleep(10);
-- end_ignore

SELECT quota_in_mb, nspsize_in_bytes from diskquota.show_fast_schema_quota_view where schema_name='ftsr';
INSERT INTO a SELECT generate_series(1,100);

DROP TABLE a;
DROP SCHEMA ftsr CASCADE;
