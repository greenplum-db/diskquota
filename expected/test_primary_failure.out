CREATE SCHEMA ftsr;
SELECT diskquota.set_schema_quota('ftsr', '1 MB');
 set_schema_quota 
------------------
 
(1 row)

SET search_path TO ftsr;
create or replace language plpythonu;
--
-- pg_ctl:
--   datadir: data directory of process to target with `pg_ctl`
--   command: commands valid for `pg_ctl`
--   command_mode: modes valid for `pg_ctl -m`  
--
create or replace function pg_ctl(datadir text, command text, command_mode text default 'immediate')
returns text as $$
    import subprocess
    if command not in ('stop', 'restart'):
        return 'Invalid command input'

    cmd = 'pg_ctl -l postmaster.log -D %s ' % datadir
    cmd = cmd + '-W -m %s %s' % (command_mode, command)

    return subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True).replace('.', '')
$$ language plpythonu;
CREATE TABLE a(i int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
INSERT INTO a SELECT generate_series(1,100);
INSERT INTO a SELECT generate_series(1,100000);
SELECT pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);
ERROR:  schema's disk space quota exceeded with name:ftsr
-- now one of primary is down
select pg_ctl((select datadir from gp_segment_configuration c where c.role='p' and c.content=0), 'stop');
        pg_ctl        
----------------------
 server shutting down+
 
(1 row)

-- switch mirror to primary
select gp_request_fts_probe_scan();
 gp_request_fts_probe_scan 
---------------------------
 t
(1 row)

-- check GPDB status
select content, preferred_role, role, status, mode from gp_segment_configuration where content = 0;
 content | preferred_role | role | status | mode 
---------+----------------+------+--------+------
       0 | p              | m    | d      | n
       0 | m              | p    | u      | n
(2 rows)

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);
ERROR:  schema's disk space quota exceeded with name:ftsr
-- increase quota
SELECT diskquota.set_schema_quota('ftsr', '200 MB');
 set_schema_quota 
------------------
 
(1 row)

-- pull up failed primary
-- start_ignore
-- end_ignore
-- check GPDB status
select content, preferred_role, role, status, mode from gp_segment_configuration where content = 0;
 content | preferred_role | role | status | mode 
---------+----------------+------+--------+------
       0 | p              | p    | u      | s
       0 | m              | m    | u      | s
(2 rows)

-- no sleep, it will take effect immediately
SELECT quota_in_mb, nspsize_in_bytes from diskquota.show_schema_quota_view where schema_name='ftsr';
 quota_in_mb | nspsize_in_bytes 
-------------+------------------
         200 |          1310720
(1 row)

INSERT INTO a SELECT generate_series(1,100);
DROP TABLE a;
DROP SCHEMA ftsr CASCADE;
NOTICE:  drop cascades to function pg_ctl(text,text,text)
